android-components/components/support/utils/src/main/java/mozilla/components/support/utils/DownloadUtils.kt

```diff
@@ -335,26 +335,30 @@ object DownloadUtils {
     /**
      * Compare the filename extension with the mime type and change it if necessary.
      */
-    private fun changeExtension(filename: String, mimeType: String?): String {
-        var extension: String? = null
-        val dotIndex = filename.lastIndexOf('.')
-
-        if (mimeType != null) {
-            val mimeTypeMap = MimeTypeMap.getSingleton()
-            // Compare the last segment of the extension against the mime type.
-            // If there's a mismatch, discard the entire extension.
-            val typeFromExt = mimeTypeMap.getMimeTypeFromExtension(filename.substringAfterLast('.'))
-            if (typeFromExt?.equals(mimeType, ignoreCase = true) != false) {
-                extension = mimeTypeMap.getExtensionFromMimeType(mimeType)?.let { ".$it" }
-                // Check if the extension needs to be changed
-                if (extension != null && filename.endsWith(extension, ignoreCase = true)) {
-                    return filename
-                }
-            }
-        }
-
-        return if (extension != null) {
-            filename.substring(0, dotIndex) + extension
+    @Suppress("ReturnCount")
+    private fun changeExtension(filename: String, providedMimeType: String?): String {
+        if (providedMimeType == null) return filename
+
+        val file = File(filename)
+        val mimeTypeMap = MimeTypeMap.getSingleton()
+        val extensionFromMimeType =
+            mimeTypeMap.getExtensionFromMimeType(providedMimeType) ?: return filename
+        val mimetypeFromFileName = mimeTypeMap.getMimeTypeFromExtension(file.extension) ?: ""
+
+        val fileHasPossibleExtension = filename.contains(extensionFromMimeType, ignoreCase = true)
+        val isFileMimetypeDifferentFromProvidedMimeType =
+            !mimetypeFromFileName.equals(
+                providedMimeType,
+                ignoreCase = true,
+            )
+
+        // Mimetypes could have multiple possible file extensions, for example: text/html could have
+        // either .htm or .html extensions, we try to rename only when there is a clear indication
+        // that the file's extension is wrong, as [getExtensionFromMimeType] returns the first possible,
+        // extension we could adding the wrong extension.
+        // More examples can be found https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
+        return if (isFileMimetypeDifferentFromProvidedMimeType && !fileHasPossibleExtension) {
+            return "${file.nameWithoutExtension}.$extensionFromMimeType"
```

Comment:
I'm not sure I'm understanding what this patch does differently in order to fix the underlying bug. It looks like they both:
1. convert the filename extension and provided mime type to an "official" mime type
2. check if the mime type matches 
    1. if they do return the original filename
    2. if they do not return the corrected filename using the provided mimetype

Reply:
If I remember correctly (When I tested 😅 ) , The main difference is that now we are checking ahead we that we already have the right extension applied, before we failed to do so, causing the [bug 1810197](https://bugzilla.mozilla.org/show_bug.cgi?id=1810197).

Reply:
Okay, I puzzled through it because I was still a little confused. As suggested in the bug, I think this line was wrong in the original:
`(typeFromExt?.equals(mimeType, ignoreCase = true) != false)`. 

That could also be written as
`(typeFromExt?.equals(mimeType, ignoreCase = true) == true)`, which means we would only look up the extension for the provided mime type when the extension already matched it. This would cause us to never replace the extension with a different type as specified by the provided mime type.

I think in your version the logic is actually handled correctly, but I am not sure why the contributor patch was reverted now. I am happy to accept that your version is doing the thing we want, but I want to just double check again that I'm not missing anything. Thanks!

Reply:
I added a comment on the contributor's [PR](https://github.com/mozilla-mobile/firefox-android/pull/1572/files#r1179593362) :)